### PR TITLE
[WIP] Server Docker Image

### PR DESCRIPTION
**Pull request by niccokunzmann**
This is work in progress. I would like to create a version of the server image which can be set up with a low skill level and only a few required options.

If you have suggestions, they are welcome. If I fail, this is a pull request which people can look at later.

TODO:

- [x] how to configure olsrd
- [ ] start the image
- [ ] connect to it with a valid installation
- [ ] create automated build or delete the section about providing the image [comment](https://github.com/wlanslovenija/tunneldigger/pull/99/files/75095f220bf03ea07845d24861d0dd9da635e121#r295940349)
- [X] clarify if OLSR is needed/nice to have/useless [comment](https://github.com/wlanslovenija/tunneldigger/pull/99#discussion_r295939518)
- [ ] add meaning of [comment](https://github.com/wlanslovenija/tunneldigger/pull/99#discussion_r295985587) in the [routing deamon](https://tunneldigger.readthedocs.io/en/latest/server.html#routing-daemon) Section
- [x] remove OLSR
- [ ] add DHCP for client and broker

### CODE REVIEW

File: broker/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+# https://github.com/OLSR/olsrd/
+#
+RUN apt-get -y install bison flex && \
+    mkdir -p /srv/olsrd && \
+    cd /srv/olsrd && \
+    wget -O- https://github.com/OLSR/olsrd/archive/v0.9.0.3.tar.gz \
+        | tar zxf - && \
+    cd * && \
+    make build_all && \
+    make install_all
```
Comment (RalfJung) [inline broker/Dockerfile]: I don't remember doing this on my servers where I run tunneldigger. Why is this needed?

File: docs/server.rst [hunk 2]
```diff
@@ -184,3 +184,45 @@ After you configured Tunneldigger, you can run the broker::
 
     cd /srv/tunneldigger
     /srv/env_tunneldigger/bin/python -m tunneldigger_broker.main /srv/tunneldigger/broker/l2tp_broker.cfg
+
+Docker Image
+------------
+
+There is a docker image available from `wlanslovenija/tunneldigger-broker
+<https://hub.docker.com/r/wlanslovenija/tunneldigger-broker>`_.
```
Comment (RalfJung) [inline docs/server.rst]: I am not sure if I am comfortable providing "official" docker images. Isn't the `docker build` below enough?

Disclaimer: I have very little docker experience, and the experience I have consists mostly of fights and frustration about how much disk space it is using.

**Comment by RalfJung** [comment]
Is there any way CI could use the docker images? Maybe even instead of the LXC stuff it is currently using?

CI looks rather messy though so don't consider this mandatory.

File: broker/Dockerfile (hunk 1 above)
Comment (niccokunzmann) [reply broker/Dockerfile]: https://tunneldigger.readthedocs.io/en/latest/server.html#routing-daemon says

> The example hook scripts require that the routing daemon (like olsrd) be configured with the Tunneldigger bridge interfaces.

I have no idea and so I added olsr.

File: docs/server.rst (hunk 2 above)
Comment (niccokunzmann) [reply docs/server.rst]: I can not answer if you are comfortable with this.
I added a TODO.

> create automated build or delete the section about providing the image

I would be able to configure the automated build as well as integrate it into the travis build. However, as this image is not of any use at this point in time, I do not work on it now.

**Comment by niccokunzmann** [comment]
> Is there any way CI could use the docker images? Maybe even instead of the LXC stuff it is currently using?

Yes. It is quite easy. However, I do not know what you expect in the build. If I knew, I would know if the image works.

For integration into the CI, we need this:
1. 3 lines activate the docker service in the image
2. build the docker file
3. run the docker file instead of the 'native' version

File: broker/Dockerfile (hunk 1 above)
Comment (RalfJung) [reply broker/Dockerfile]: That's a very environment-specific part of the configuration. To my knowledge, most Freifunk Communities use Batman instead of OLSR, and would have to configure that.

I don't think it is possible to provide something in Docker that would make this reusable. But including OLSR here makes little sense IMO. People can derive their own Docker images that contains tunneldigger + whatever they need.

I can't speak for other communities, but in Freifunk Saar I don't think it would be possible to isolate tunneldigger from the rest of the network stuff that is going on on a gateway. All the routing table magic and bridge configuration and batman needs to happen in the same network namespace, so it cannot be easily split across containers.

File: docs/server.rst (hunk 2 above)
Comment (RalfJung) [reply docs/server.rst]: Yeah, I think the image is more useful as "machine-readable installation instructions" than for direct (re-)use. So it probably makes more sense not to have it on the hub.

**Comment by RalfJung** [comment]
> However, I do not know what you expect in the build.

Well, it should carry our the same tests as is does now, just using the Docker image for the server instead of the LXC container that it uses currently.

I am not very familiar with the CI though, I entirely inherited that. ;)

File: broker/Dockerfile (hunk 1 above)
Comment (niccokunzmann) [reply broker/Dockerfile]: Oh. Thanks for clarifying this. 
I checked that point above. Other images can build on top of the provided tunnel digger image.

File: docs/server.rst (hunk 2 above)
Comment (niccokunzmann) [reply docs/server.rst]: I would create a new issue for hosting it. For me, it would be useful as I can build another image on top of it. However, I will create the issue when I succeed with this image.

**Comment by niccokunzmann** [comment]
I integrated the docker build. 37cc7d7 
What I struggle with: Tests and container building are interwoven: 

https://github.com/wlanslovenija/tunneldigger/blob/189ed38866de95d029f2f44b25330c399be81436/tests/test_nose.py#L28
https://github.com/wlanslovenija/tunneldigger/blob/189ed38866de95d029f2f44b25330c399be81436/tests/tunneldigger.py#L191

> Is there any way CI could use the docker images? Maybe even instead of the LXC stuff it is currently using?

Docker [works on top of LXC](https://unix.stackexchange.com/questions/254956/what-is-the-difference-between-docker-lxd-and-lxc). So, I do not see a need to replace this.

For me the next step is to understand what I can do to run tunneldigger in general. I do not know how to run it, how it works, how I can see if it works. If I know that, I know how to test the docker image and vice versa.

**Comment by RalfJung** [comment]
> Docker works on top of LXC. So, I do not see a need to replace this.

Sure, but currently the test scripts interact directly with the LXC API (in [this file](https://github.com/wlanslovenija/tunneldigger/blob/c852397209f63416a19a9a984e4e9cebcbea049b/tests/tunneldigger.py)), while Docker would like to do that itself and then presents its own API to users.

> For me the next step is to understand what I can do to run tunneldigger in general. I do not know how to run it, how it works, how I can see if it works. If I know that, I know how to test the docker image and vice versa.

What the CI currently does is to spawn a tunneldigger server and client, make the client connect to the server -- this establishes an l2tp tunnel between the to. Then it sends some pings through that connection to see if the tunnel is actually working. It also seems to set up a lighthttpd to test that data can flow through the tunnel. It also does some stuff to test MTU-related logic, which I understand less.

Your docker image is for the server, so the client part of the existing tests could likely remain unchanged. The point of the server part is to ultimately run [this script](https://github.com/wlanslovenija/tunneldigger/blob/c852397209f63416a19a9a984e4e9cebcbea049b/tests/run_server.sh) that starts the actual server process. Before that, the preparation of the tunneldigger server config happens in [this script](https://github.com/wlanslovenija/tunneldigger/blob/master/tests/prepare_server.sh). Both of these get called by the [test "driver"](https://github.com/wlanslovenija/tunneldigger/blob/c852397209f63416a19a9a984e4e9cebcbea049b/tests/tunneldigger.py) that also does all the LXC setup.

**Comment by RalfJung** [comment]
How to run the server manually is [described in the docs](https://tunneldigger.readthedocs.io/en/latest/server.html#running). What we do on our gateways is we run the following script:
```
#!/bin/sh
set -e

cd /opt/
virtualenv tunneldigger
cd tunneldigger
. bin/activate

cd broker
python setup.py install
exec python -m tunneldigger_broker.main /opt/tunneldigger/l2tp_broker.cfg
```
Most of that is just the usual python virtualenv stuff. Then the last two lines install tunneldigger dependencies, and finally run the actual server.

For the config file, there is an example [here](https://github.com/wlanslovenija/tunneldigger/blob/master/broker/l2tp_broker.cfg.example), from which the existing test suite creates one via `sed`:
```
cp /srv/tunneldigger/broker/l2tp_broker.cfg.example /srv/tunneldigger/broker/l2tp_broker.cfg || true
sed -i "s/127.0.0.1/$IP/g" /srv/tunneldigger/broker/l2tp_broker.cfg
sed -i "s/^interface=.*/interface=eth1/g" /srv/tunneldigger/broker/l2tp_broker.cfg
```

That should at least get the tunneldigger daemon running.


The client takes the following options:
```
$ ./tunneldigger --help
./tunneldigger: invalid option -- '-'
ERROR: Invalid option ?!
usage: ./tunneldigger [options]
       -h            this text
       -f            don't daemonize into background
       -u uuid       set UUID string
       -l ip         local IP address to bind to (default 0.0.0.0)
       -b host:port  broker hostname and port (can be specified multiple times)
       -i iface      tunnel interface name
       -I iface      force client to bind tunnel socket to a specific interface
       -s hook       hook script
       -t id         local tunnel id (default 1)
       -L limit      request broker to set downstream bandwidth limit (in kbps)
       -a            select broker based on use
       -g            select first available broker to connect to (default)
       -r            select a random broker
```
The test suite seems to start it as `tunneldigger -u foobar -i l2tp0 -t 2 -L 102400 -s /testing/hook_client.sh -f -b SERVERADDR:SERVERPORT`.

If that works, you should see some logging talking about a successful connection, and there should be a new `l2tp*` network interface on the server and one on the client that are connected to each other and that can transport network data.

**Comment by niccokunzmann** [comment]
Celebration :fireworks: The current image outputs this:

```
Chain PREROUTING (policy ACCEPT)
target     prot opt source               destination         

Chain INPUT (policy ACCEPT)
target     prot opt source               destination         

Chain OUTPUT (policy ACCEPT)
target     prot opt source               destination         

Chain POSTROUTING (policy ACCEPT)
target     prot opt source               destination         
[INFO/tunneldigger.broker] Initializing the tunneldigger broker.
[INFO/tunneldigger.broker] Registered script '/srv/tunneldigger/tunneldigger/broker/scripts/setup_interface.sh' for hook 'session.up'.
[INFO/tunneldigger.broker] Registered script '/srv/tunneldigger/tunneldigger/broker/scripts/teardown_interface.sh' for hook 'session.pre-down'.
[INFO/tunneldigger.broker] Registered script '/srv/tunneldigger/tunneldigger/broker/scripts/mtu_changed.sh' for hook 'session.mtu-changed'.
[INFO/tunneldigger.broker] Maximum number of tunnels is 1024.
[INFO/tunneldigger.broker] Tunnel identifier base is 100.
[INFO/tunneldigger.broker] Tunnel port base is 20000.
[INFO/tunneldigger.broker] Namespace is default.
[INFO/tunneldigger.broker] Listening on 172.17.0.3:53.
[INFO/tunneldigger.broker] Listening on 172.17.0.3:123.
[INFO/tunneldigger.broker] Listening on 172.17.0.3:8942.
[INFO/tunneldigger.broker] Broker initialized.
```

**Comment by niccokunzmann** [comment]
So, this is the progress: client and broker communicate but are unable to establish the connection.

```
Chain PREROUTING (policy ACCEPT)
target     prot opt source               destination         

Chain INPUT (policy ACCEPT)
target     prot opt source               destination         

Chain OUTPUT (policy ACCEPT)
target     prot opt source               destination         

Chain POSTROUTING (policy ACCEPT)
target     prot opt source               destination         
[INFO/tunneldigger.broker] Initializing the tunneldigger broker.
[INFO/tunneldigger.broker] Registered script '/srv/tunneldigger/tunneldigger/broker/scripts/setup_interface.sh' for hook 'session.up'.
[INFO/tunneldigger.broker] Registered script '/srv/tunneldigger/tunneldigger/broker/scripts/teardown_interface.sh' for hook 'session.pre-down'.
[INFO/tunneldigger.broker] Registered script '/srv/tunneldigger/tunneldigger/broker/scripts/mtu_changed.sh' for hook 'session.mtu-changed'.
[INFO/tunneldigger.broker] Maximum number of tunnels is 1024.
[INFO/tunneldigger.broker] Tunnel identifier base is 100.
[INFO/tunneldigger.broker] Tunnel port base is 20000.
[INFO/tunneldigger.broker] Namespace is default.
[INFO/tunneldigger.broker] Listening on 172.17.0.3:53.
[INFO/tunneldigger.broker] Listening on 172.17.0.3:123.
[INFO/tunneldigger.broker] Listening on 172.17.0.3:8942.
[INFO/tunneldigger.broker] Broker initialized.
[INFO/tunneldigger.broker] Creating tunnel 172.17.0.1:48642 (+ixqShc) with id 100.
[INFO/tunneldigger.tunnel] Set tunnel 100 MTU to 1446.
[INFO/tunneldigger.hooks] Running hook 'session.up' via script '/srv/tunneldigger/tunneldigger/broker/scripts/setup_interface.sh 100 100 l2tp100-100 1446 172.17.0.1 48642 20100 +ixqShc 53'.
[INFO/tunneldigger.hooks] (session.up/324) Error: argument "mesh" is wrong: invalid table ID
[INFO/tunneldigger.hooks] (session.up/324) Error: argument "mesh" is wrong: invalid table ID
[INFO/tunneldigger.hooks] (session.up/324) modprobe: ERROR: ../libkmod/libkmod.c:586 kmod_search_moddep() could not open moddep file '/lib/modules/4.14.74-1.pvops.qubes.x86_64/modules.dep.bin'
[INFO/tunneldigger.hooks] (session.up/324) modprobe: FATAL: Module ebtables not found in directory /lib/modules/4.14.74-1.pvops.qubes.x86_64
[INFO/tunneldigger.hooks] (session.up/324) The kernel doesn't support the ebtables 'filter' table.
[INFO/tunneldigger.broker] Rejecting tunnel 172.17.0.1:48468 (+ixqShc) due to rate limiting
[WARNING/tunneldigger.protocol] Failed to create tunnel (+ixqShc) while processing prepare request.
[INFO/tunneldigger.broker] Rejecting tunnel 172.17.0.1:52147 (+ixqShc) due to rate limiting
[WARNING/tunneldigger.protocol] Failed to create tunnel (+ixqShc) while processing prepare request.
[INFO/tunneldigger.broker] Rejecting tunnel 172.17.0.1:53654 (+ixqShc) due to rate limiting
[WARNING/tunneldigger.protocol] Failed to create tunnel (+ixqShc) while processing prepare request.
[INFO/tunneldigger.broker] Creating tunnel 172.17.0.1:41576 (+ixqShc) with id 101.
[INFO/tunneldigger.tunnel] Set tunnel 101 MTU to 1446.
[INFO/tunneldigger.hooks] Running hook 'session.up' via script '/srv/tunneldigger/tunneldigger/broker/scripts/setup_interface.sh 101 101 l2tp101-101 1446 172.17.0.1 41576 20101 +ixqShc 53'.
[INFO/tunneldigger.broker] Rejecting tunnel 172.17.0.1:33781 (+ixqShc) due to rate limiting
[WARNING/tunneldigger.protocol] Failed to create tunnel (+ixqShc) while processing prepare request.
[INFO/tunneldigger.broker] Rejecting tunnel 172.17.0.1:45424 (+ixqShc) due to rate limiting
[WARNING/tunneldigger.protocol] Failed to create tunnel (+ixqShc) while processing prepare request.
[INFO/tunneldigger.broker] Rejecting tunnel 172.17.0.1:41307 (+ixqShc) due to rate limiting
[WARNING/tunneldigger.protocol] Failed to create tunnel (+ixqShc) while processing prepare request.
[INFO/tunneldigger.broker] Creating tunnel 172.17.0.1:47034 (+ixqShc) with id 102.
[INFO/tunneldigger.tunnel] Set tunnel 102 MTU to 1446.
[INFO/tunneldigger.hooks] Running hook 'session.up' via script '/srv/tunneldigger/tunneldigger/broker/scripts/setup_interface.sh 102 102 l2tp102-102 1446 172.17.0.1 47034 20102 +ixqShc 53'.
[INFO/tunneldigger.broker] Rejecting tunnel 172.17.0.1:38579 (+ixqShc) due to rate limiting
[WARNING/tunneldigger.protocol] Failed to create tunnel (+ixqShc) while processing prepare request.
```

client:

```
td-client: Performing broker selection...
td-client: Broker usage of 172.17.0.3:53: 127
td-client: Selected 172.17.0.3:53 as the best broker.
td-client: Received error response from broker!
td-client: Connection to 172.17.0.3 lost.
td-client: Performing broker selection...
td-client: Broker usage of 172.17.0.3:53: 127
td-client: Selected 172.17.0.3:53 as the best broker.
td-client: Unable to create local L2TP tunnel!
td-client: Connection to 172.17.0.3 lost.
td-client: Performing broker selection...
td-client: Broker usage of 172.17.0.3:53: 191
td-client: Selected 172.17.0.3:53 as the best broker.
td-client: Received error response from broker!
td-client: Connection to 172.17.0.3 lost.
td-client: Performing broker selection...
td-client: Broker usage of 172.17.0.3:53: 191
td-client: Selected 172.17.0.3:53 as the best broker.
```

**Comment by niccokunzmann** [comment]
The output now is better:
broker:
```
Chain PREROUTING (policy ACCEPT)
target     prot opt source               destination         

Chain INPUT (policy ACCEPT)
target     prot opt source               destination         

Chain OUTPUT (policy ACCEPT)
target     prot opt source               destination         

Chain POSTROUTING (policy ACCEPT)
target     prot opt source               destination         
[INFO/tunneldigger.broker] Initializing the tunneldigger broker.
[INFO/tunneldigger.broker] Registered script '/srv/tunneldigger/tunneldigger/broker/scripts/setup_interface.sh' for hook 'session.up'.
[INFO/tunneldigger.broker] Registered script '/srv/tunneldigger/tunneldigger/broker/scripts/teardown_interface.sh' for hook 'session.pre-down'.
[INFO/tunneldigger.broker] Registered script '/srv/tunneldigger/tunneldigger/broker/scripts/mtu_changed.sh' for hook 'session.mtu-changed'.
[INFO/tunneldigger.broker] Maximum number of tunnels is 1024.
[INFO/tunneldigger.broker] Tunnel identifier base is 100.
[INFO/tunneldigger.broker] Tunnel port base is 20000.
[INFO/tunneldigger.broker] Namespace is default.
[INFO/tunneldigger.broker] Listening on 172.17.0.2:53.
[INFO/tunneldigger.broker] Listening on 172.17.0.2:123.
[INFO/tunneldigger.broker] Listening on 172.17.0.2:8942.
[INFO/tunneldigger.broker] Broker initialized.
[INFO/tunneldigger.broker] Creating tunnel (EJiPw+s) with id 100.
[INFO/tunneldigger.tunnel] Set tunnel 100 MTU to 1446.
[INFO/tunneldigger.hooks] Running hook 'session.up' via script '/srv/tunneldigger/tunneldigger/broker/scripts/setup_interface.sh'.
[INFO/tunneldigger.hooks] (session.up/35) RTNETLINK answers: No such file or directory
```

Client:
```
/usr/local/bin/tunneldigger -f -i l2tp0 -s /srv/tunneldigger/client/hook.sh -u EJiPw+s -b 172.17.0.2:123
td-client: Performing broker selection...
td-client: Broker usage of 172.17.0.2:123: 0
td-client: Selected 172.17.0.2:123 as the best broker.
td-client: Tunnel successfully established.
td-client: Setting MTU to 1446
td-client: Setting MTU to 1446
```

Client and broker can not ping eachother.

Broker routes:
```
root@c4817355f257:/opt/src# route
Kernel IP routing table
Destination     Gateway         Genmask         Flags Metric Ref    Use Iface
default         172.17.0.1      0.0.0.0         UG    0      0        0 eth0
10.254.0.0      0.0.0.0         255.255.0.0     U     0      0        0 digger1446
172.17.0.0      0.0.0.0         255.255.0.0     U     0      0        0 eth0
192.168.254.0   0.0.0.0         255.255.255.0   U     0      0        0 br0
```

I wonder: I would like the broker to give ip addresses to the clients. How this can be achieved?

**Comment by niccokunzmann** [comment]
As shown in https://github.com/wlanslovenija/tunneldigger/pull/104#issuecomment-508924660, both docker images can connect to each other. What is left from my perspective is to add dhcp in the interfaces so that the ip addresses are configured automatically. This is because I assume this to be the main use-case. I do not know it though. Do you know more about this?

**Comment by RalfJung** [comment]
(I'm super busy for about one more week, then I'll start to work on my backlog, including this PR.)

**Comment by RalfJung** [comment]
> What is left from my perspective is to add dhcp in the interfaces so that the ip addresses are configured automatically. This is because I assume this to be the main use-case. I do not know it though. Do you know more about this?

Actually in our usecase we do not run IP over that interface at all. We run batman, a layer 2 mesh protocol, and then we run IP *in there*.

So as I said before, I do not think these docker images will be useful for production use. Do you use them for that?

But they might still be useful for documentation purpose, IMO.

File: broker/docker/add_to_ip [hunk 3]
```diff
@@ -0,0 +1,9 @@
+#!/usr/bin/python
+
+import sys
+ip = sys.argv[1]
+ip_int = sum(int(part) * 256 ** (3 - i) for i, part in enumerate(ip.split(".")))
+add = int(sys.argv[2])
+new_ip = add + ip_int
+print("%d.%d.%d.%d" % tuple([0xff & (new_ip / 256 ** (3 - i)) for i in range(4)]))
```
Comment (RalfJung) [inline broker/docker/add_to_ip]: What is the purpose of this script (and `ip_net_to_net_mask`)? They should both have comments explaining that.

File: broker/docker/docker-compose.yml [hunk 4]
```diff
@@ -0,0 +1,5 @@
+version: "2"
+
+services:
+    tunneldigger:
+        build: .
```
Comment (RalfJung) [inline broker/docker/docker-compose.yml]: Does docker-compose really simplify anything here?

(Note: I have basically no experience with docker-compose directly, but suffered a lot of pain from the Discourse docker stuff which I think uses docker-compose.)

File: broker/docker/start.sh [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+
+#
+# activate modules
+#
+
+#modprobe l2tp_netlink
+#modprobe l2tp_ip
+#modprobe l2tp_eth
+#modprobe nf_conntrack
+#modprobe nf_conntrack_netlink
```
Comment (RalfJung) [inline broker/docker/start.sh]: Why is this commented out?

File: broker/docker/start.sh [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+
+#
+# configure bridge, see
+# https://github.com/wlanslovenija/tunneldigger/blob/master/tests/prepare_server.sh
+#
+export IP_NET="${IP_NET:-10.254.0.1/16}"
+# we never use the br0 interface
+#brctl addbr br0
+#ip a a $IP_NET dev br0
+#ip l s br0 up
```
Comment (RalfJung) [inline broker/docker/start.sh]: Why is this code not removed?

File: broker/scripts/bridge_functions.sh [hunk 7]
```diff
@@ -13,7 +15,7 @@ ensure_bridge()
     # Bridge did not exist before, we have to initialize it
     ip link set dev $brname up
     # TODO The IP address should probably not be hardcoded here?
```
Comment (RalfJung) [inline broker/scripts/bridge_functions.sh]: This TODO seems outdated now?

File: docs/server.rst [hunk 8]
```diff
@@ -184,3 +184,48 @@ After you configured Tunneldigger, you can run the broker::
...
+               -e interface=lo      -e max_tunnels=1024 \
+               -e port_base=20000   -e tunnel_id_base=100 \
+               -e namespace=default -e connection_rate_limit=10 \
+               -e pmtu=0            -e verbosity=DEBUG \
+               -e log_ip_addresses=false --privileged \
+               -it wlanslovenija/tunneldigger-broker
+
+Instead of configuring the environment variables one by one
+using the ``-e`` option, you can also use the ``--env-file``
+option.
```
Comment (RalfJung) [inline docs/server.rst]: Same question as for the client image: when is this server image *useful*? If the answer is "mostly for documentation and testing", that's okay, but then it should say so.

File: .travis.yml [hunk 9]
```diff
@@ -19,6 +21,7 @@ install:
   - sudo iptables -t nat -L >/dev/null
   - sudo iptables -t filter -P FORWARD ACCEPT
   - sudo /usr/bin/python3 ./tests/tunneldigger.py --setup
+  - ( cd broker; docker build --tag wlanslovenija/tunneldigger-broker . )
```
Comment (RalfJung) [inline .travis.yml]: So this just checks if the docker image builds, right, but does not run it?

If yes, no need to do that in *each* Travis job.

**Comment by RalfJung** [comment]
Closing due to lack of activity. If you want to pick this up again @niccokunzmann, feel free to reopen!
